### PR TITLE
Freertos initial

### PR DESCRIPTION
**Pull request by rbudai98**
* FreeRTOS generic makefile added
* Current build commands need: `FREERTOS=y` for enabling the library and `FREERTOS_KERNEL=/path/to/Kernel` for specifying the path towards freeRTOS Kernel (https://github.com/FreeRTOS/FreeRTOS-Kernel)
* FreeRTOS.config file is linked from project folder
* Current implementation has the ARM4F CMSIS file linked for max devices, in future it will need generalizations
* (example project in different PR)

### CODE REVIEW

**Comment by rbudai98** [comment]
v2:
* added mutex includes to projects 
* corrected doxygen for platform/freertos files

File: include/no_os_eeprom.h [hunk 1]
```diff
@@ -79,6 +81,8 @@ struct no_os_eeprom_desc {
 	const struct no_os_eeprom_platform_ops *platform_ops;
 	/** EEPROM extra parameters (device specific) */
 	void *extra;
+	/** EEPROM mutex*/
+	void 		*mutex;
```
Comment (buha) [inline include/no_os_eeprom.h:85]: I would put the mutex as the first member of this structure to be able to lock the resource by a simple cast. It might be useful in the future.

File: drivers/platform/freeRTOS/freertos_semaphore.c [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * @brief Take token from semaphore.
+ * semaphore - Pointer toward the semaphore.
+ * @return None.
+ */
+__attribute__((weak)) inline void no_os_semaphore_take(void *semaphore)
+{
+	xSemaphoreTake(semaphore,
+		       freeRTOS_WAIT_TIME); //replace with define from freertos config file
```
Comment (buha) [inline drivers/platform/freeRTOS/freertos_semaphore.c:62]: you can remove the comment I think

File: tools/scripts/freertos.mk [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+#Link heap implementation
+RTOS_SRCS = $(FREERTOS_KERNEL)/portable/MemMang/heap_2.c
+
+RTOS_SRCS += $(FREERTOS_KERNEL)/tasks.c \
+        $(FREERTOS_KERNEL)/list.c \
+        $(FREERTOS_KERNEL)/queue.c \
+        $(FREERTOS_KERNEL)/timers.c 
+
+RTOS_SRCS += $(FREERTOS_KERNEL)/portable/GCC/ARM_CM4F/port.c
```
Comment (buha) [inline tools/scripts/freertos.mk:31]: At some point in the future it would be nice if this was selected automatically based on processor because as it is now, it will certainly cause troubles when attempting a build with FreeRTOS for CM0 or CM33 etc.

Unfortunately there is no trivial way to do this, just by string manipulation/regex.

File: drivers/api/no_os_eeprom.c [hunk 4]
```diff
@@ -121,5 +127,9 @@ int32_t no_os_eeprom_read(struct no_os_eeprom_desc *desc, uint32_t address,
 	if (!desc->platform_ops->read)
 		return -ENOSYS;
 
+	no_os_mutex_lock(desc->mutex);
+
 	return desc->platform_ops->read(desc, address, data, bytes);
+
+	no_os_mutex_unlock(desc->mutex);
```
Comment (CiprianRegus) [inline drivers/api/no_os_eeprom.c:134]: this will cause the mutex to remain locked. Return after unlocking.

File: drivers/api/no_os_i2c.c [hunk 5]
```diff
@@ -105,8 +107,12 @@ int32_t no_os_i2c_write(struct no_os_i2c_desc *desc,
 	if (!desc->platform_ops->i2c_ops_write)
 		return -ENOSYS;
 
-	return desc->platform_ops->i2c_ops_write(desc, data, bytes_number,
+	no_os_mutex_lock(desc->mutex);
+	int32_t ret = desc->platform_ops->i2c_ops_write(desc, data, bytes_number,
```
Comment (CiprianRegus) [inline drivers/api/no_os_i2c.c:111]: please declare `ret` at the start of the function. There are some other places where this applies.

File: drivers/api/no_os_spi.c [hunk 6]
```diff
@@ -125,14 +132,21 @@ int32_t no_os_spi_transfer(struct no_os_spi_desc *desc,
...
 					       msgs[i].bytes_number);
-		if (NO_OS_IS_ERR_VALUE(ret))
+		if (NO_OS_IS_ERR_VALUE(ret)) {
+			no_os_mutex_unlock(desc->mutex);
 			return ret;
+		}
 	}
 
+	no_os_mutex_unlock(desc->mutex);
 	return 0;
```
Comment (CiprianRegus) [inline drivers/api/no_os_spi.c:151]: this may be simplified by using `goto` statements as such:

```
for(...) {
ret = no_os_spi_write_and_read(desc, msgs[i].rx_buff,
					       msgs[i].bytes_number);
if (NO_OS_IS_ERR_VALUE(ret))
        goto out;
}

out:
        no_os_mutex_unlock(desc->mutex);

return ret;
```

File: include/no_os_mdio.h [hunk 7]
```diff
@@ -74,6 +74,8 @@ struct no_os_mdio_init_param {
 	struct no_os_mdio_ops *ops;
 	/** Platform or implementation specific parameters. */
 	void *extra;
+	/** MDIO mutex*/
+	void 		*mutex;
```
Comment (CiprianRegus) [inline include/no_os_mdio.h:78]: I think there may be the case where the user declares more than 1 descriptor for a peripheral. By placing the mutex in here, it won't be shared between descriptors, which might cause issues if 2 descriptors are used in 2 different threads. One possible solution would be to declare it as `static` in the corresponding source files (`no_os_mdio.c` in this case).

Comment (rbudai98) [reply include/no_os_mdio.h:78]: In general it is correct. However if we switch to `static` declaration it won't be possible to declare for instance two different SPI descriptors, that might have different ports and might be used simultaneously.

File: drivers/api/no_os_i2c.c [hunk 8]
```diff
@@ -130,6 +136,10 @@ int32_t no_os_i2c_read(struct no_os_i2c_desc *desc,
 	if (!desc->platform_ops->i2c_ops_read)
 		return -ENOSYS;
 
-	return desc->platform_ops->i2c_ops_read(desc, data, bytes_number,
-						stop_bit);
+	no_os_mutex_lock(desc->mutex);
+	int32_t ret = desc->platform_ops->i2c_ops_read(desc, data, bytes_number,
```
Comment (amiclaus) [inline drivers/api/no_os_i2c.c:140]: variable declaration at the start of the function.

File: drivers/api/no_os_mdio.c [hunk 9]
```diff
@@ -101,7 +104,11 @@ int no_os_mdio_write(struct no_os_mdio_desc *desc, uint32_t reg, uint16_t val)
 	if (!desc->ops->write)
 		return -ENOSYS;
 
-	return desc->ops->write(desc, reg, val);
+	no_os_mutex_lock(desc->mutex);
+	int ret = desc->ops->write(desc, reg, val);
```
Comment (amiclaus) [inline drivers/api/no_os_mdio.c:108]: ditto.

File: drivers/api/no_os_mdio.c [hunk 10]
```diff
@@ -121,5 +128,9 @@ int no_os_mdio_read(struct no_os_mdio_desc *desc, uint32_t reg, uint16_t *val)
 	if (!desc->ops->read)
 		return -ENOSYS;
 
-	return desc->ops->read(desc, reg, val);
+	no_os_mutex_lock(desc->mutex);
+	int ret = desc->ops->read(desc, reg, val);
```
Comment (amiclaus) [inline drivers/api/no_os_mdio.c:132]: ditto

File: drivers/api/no_os_pwm.c [hunk 11]
```diff
@@ -97,7 +100,9 @@ int32_t no_os_pwm_enable(struct no_os_pwm_desc *desc)
 	if (!desc->platform_ops->pwm_ops_enable)
 		return -ENOSYS;
 
+	no_os_mutex_lock(desc->mutex);
 	return desc->platform_ops->pwm_ops_enable(desc);
+
```
Comment (amiclaus) [inline drivers/api/no_os_pwm.c:105]: unrelated change.

File: drivers/api/no_os_spi.c [hunk 12]
```diff
@@ -41,6 +41,8 @@
 #include "no_os_spi.h"
 #include <stdlib.h>
 #include "no_os_error.h"
+#include "no_os_mutex.h"
+
```
Comment (amiclaus) [inline drivers/api/no_os_spi.c:45]: unrelated change.

File: drivers/api/no_os_spi.c [hunk 13]
```diff
@@ -102,7 +105,11 @@ int32_t no_os_spi_write_and_read(struct no_os_spi_desc *desc,
 	if (!desc->platform_ops->write_and_read)
 		return -ENOSYS;
 
-	return desc->platform_ops->write_and_read(desc, data, bytes_number);
+	no_os_mutex_lock(desc->mutex);
+	int32_t ret =  desc->platform_ops->write_and_read(desc, data, bytes_number);
```
Comment (amiclaus) [inline drivers/api/no_os_spi.c:109]: ditto.

File: drivers/api/no_os_spi.c [hunk 14]
```diff
@@ -125,14 +132,21 @@ int32_t no_os_spi_transfer(struct no_os_spi_desc *desc,
 	if (desc->platform_ops->transfer)
 		return desc->platform_ops->transfer(desc, msgs, len);
 
+	no_os_mutex_lock(desc->mutex);
+
 	for (i = 0; i < len; i++) {
-		if (msgs[i].rx_buff != msgs[i].tx_buff || !msgs[i].tx_buff)
+		if (msgs[i].rx_buff != msgs[i].tx_buff || !msgs[i].tx_buff) {
+			no_os_mutex_unlock(desc->mutex);
```
Comment (amiclaus) [inline drivers/api/no_os_spi.c:139]: this could be done smoother with a conditional jump.

File: drivers/api/no_os_uart.c [hunk 15]
```diff
@@ -117,7 +119,11 @@ int32_t no_os_uart_read(struct no_os_uart_desc *desc,
 	if (!desc->platform_ops->read)
 		return -ENOSYS;
 
-	return desc->platform_ops->read(desc, data, bytes_number);
+	no_os_mutex_lock(desc->mutex);
+	int32_t ret = desc->platform_ops->read(desc, data, bytes_number);
```
Comment (amiclaus) [inline drivers/api/no_os_uart.c:123]: ditto

File: drivers/api/no_os_uart.c [hunk 16]
```diff
@@ -137,7 +143,11 @@ int32_t no_os_uart_write(struct no_os_uart_desc *desc,
 	if (!desc->platform_ops->write)
 		return -ENOSYS;
 
-	return desc->platform_ops->write(desc, data, bytes_number);
+	no_os_mutex_lock(desc->mutex);
+	int32_t ret = desc->platform_ops->write(desc, data, bytes_number);
```
Comment (amiclaus) [inline drivers/api/no_os_uart.c:147]: ditto.

File: drivers/api/no_os_uart.c [hunk 17]
```diff
@@ -157,7 +167,11 @@ int32_t no_os_uart_read_nonblocking(struct no_os_uart_desc *desc,
 	if (!desc->platform_ops->read_nonblocking)
 		return -ENOSYS;
 
-	return desc->platform_ops->read_nonblocking(desc, data, bytes_number);
+	no_os_mutex_lock(desc->mutex);
+	int32_t ret = desc->platform_ops->read_nonblocking(desc, data, bytes_number);
```
Comment (amiclaus) [inline drivers/api/no_os_uart.c:171]: ditto

File: drivers/api/no_os_uart.c [hunk 18]
```diff
@@ -177,7 +191,11 @@ int32_t no_os_uart_write_nonblocking(struct no_os_uart_desc *desc,
 	if (!desc->platform_ops->write_nonblocking)
 		return -ENOSYS;
 
-	return desc->platform_ops->write_nonblocking(desc, data, bytes_number);
+	no_os_mutex_lock(desc->mutex);
+	int32_t ret = desc->platform_ops->write_nonblocking(desc, data, bytes_number);
```
Comment (amiclaus) [inline drivers/api/no_os_uart.c:195]: ditto

File: drivers/platform/freeRTOS/freertos_delay.c [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ * @brief Generate microseconds delay.
+ * @param usecs - Delay in microseconds.
+ * @return None.
+ */
+
+void no_os_udelay(uint32_t usecs)
+{
+	vTaskDelay(((usecs/1000000) * configTICK_RATE_HZ));
```
Comment (amiclaus) [inline drivers/platform/freeRTOS/freertos_delay.c:52]: nit: spaces before and after `/`

File: drivers/platform/freeRTOS/freertos_delay.c [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+/**
+ * @brief Generate miliseconds delay.
+ * @param msecs - Delay in miliseconds.
+ * @return None.
+ */
+void no_os_mdelay(uint32_t msecs)
+{
+	vTaskDelay(((msecs/1000) * configTICK_RATE_HZ));
```
Comment (amiclaus) [inline drivers/platform/freeRTOS/freertos_delay.c:62]: ditto: spaces before and after `/`

File: drivers/platform/freeRTOS/freertos_delay.c [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+/**
+ * @brief Get current time.
+ * @return Current time structure from system start (seconds, microseconds).
+ */
+struct no_os_time no_os_get_time(void)
+{
+	unsigned long long Xtime_Global;
+	float fractional_part = 0;
```
Comment (amiclaus) [inline drivers/platform/freeRTOS/freertos_delay.c:72]: can we avoid `float`?

File: drivers/platform/freeRTOS/freertos_semaphore.c [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * @brief Remove semaphore.
+ * semaphore - Pointer toward the semaphore.
+ * @return None.
+ */
+__attribute__((weak)) inline void no_os_semaphore_remove(void *semaphore)
+{
+	no_os_free(semaphore);
+}
+
```
Comment (amiclaus) [inline drivers/platform/freeRTOS/freertos_semaphore.c:84]: nit: remove line.

File: projects/ad738x_fmcz/src.mk [hunk 23]
```diff
@@ -32,4 +33,6 @@ INCS +=	$(INCLUDE)/no_os_axi_io.h \
 	$(INCLUDE)/no_os_delay.h \
 	$(INCLUDE)/no_os_print_log.h \
 	$(INCLUDE)/no_os_util.h \
-	$(INCLUDE)/no_os_alloc.h
+	$(INCLUDE)/no_os_alloc.h \
+	$(INCLUDE)/no_os_mutex.h 
+
```
Comment (amiclaus) [inline projects/ad738x_fmcz/src.mk:38]: nit: unrelated change.

File: projects/adxrs290-pmdz/src.mk [hunk 24]
```diff
@@ -27,14 +27,16 @@ INCS += $(INCLUDE)/no_os_delay.h        \
...
         $(INCLUDE)/no_os_units.h        \
-        $(INCLUDE)/no_os_alloc.h
+        $(INCLUDE)/no_os_alloc.h        \
+        $(INCLUDE)/no_os_mutex.h
 
 SRCS += $(DRIVERS)/api/no_os_gpio.c     \
         $(NO-OS)/util/no_os_lf256fifo.c \
         $(DRIVERS)/api/no_os_irq.c      \
-         $(DRIVERS)/api/no_os_timer.c   \
+        $(DRIVERS)/api/no_os_timer.c    \
```
Comment (amiclaus) [inline projects/adxrs290-pmdz/src.mk:36]: nit: unrelated change.

File: projects/adt75/src.mk [hunk 25]
```diff
@@ -23,7 +23,8 @@ INCS += $(INCLUDE)/no_os_delay.h     		\
 		$(INCLUDE)/no_os_lf256fifo.h 	\
 		$(INCLUDE)/no_os_util.h 	\
 		$(INCLUDE)/no_os_units.h        \
-		$(INCLUDE)/no_os_alloc.h
+		$(INCLUDE)/no_os_alloc.h		\
+        $(INCLUDE)/no_os_mutex.h
```
Comment (amiclaus) [inline projects/adt75/src.mk:27]: odd indentation.

File: projects/eval-adis/src.mk [hunk 26]
```diff
@@ -24,7 +24,8 @@ INCS += $(INCLUDE)/no_os_delay.h     \
 		$(INCLUDE)/no_os_lf256fifo.h \
 		$(INCLUDE)/no_os_util.h \
 		$(INCLUDE)/no_os_units.h \
-		$(INCLUDE)/no_os_alloc.h
+		$(INCLUDE)/no_os_alloc.h \
+        $(INCLUDE)/no_os_mutex.h
```
Comment (amiclaus) [inline projects/eval-adis/src.mk:28]: ditto: odd indentation.

File: projects/eval-adxl313z/src.mk [hunk 27]
```diff
@@ -24,7 +24,8 @@ INCS += $(INCLUDE)/no_os_delay.h     \
 		$(INCLUDE)/no_os_util.h \
 		$(INCLUDE)/no_os_units.h \
 		$(INCLUDE)/no_os_init.h \
-		$(INCLUDE)/no_os_alloc.h
+		$(INCLUDE)/no_os_alloc.h \
+        $(INCLUDE)/no_os_mutex.h
```
Comment (amiclaus) [inline projects/eval-adxl313z/src.mk:28]: ditto.

File: projects/eval-adxl313z/src.mk [hunk 28]
```diff
@@ -34,7 +35,8 @@ SRCS += $(DRIVERS)/api/no_os_gpio.c \
 		$(NO-OS)/util/no_os_list.c \
 		$(DRIVERS)/api/no_os_uart.c \
 		$(NO-OS)/util/no_os_util.c \
-		$(NO-OS)/util/no_os_alloc.c
+		$(NO-OS)/util/no_os_alloc.c \
+        $(NO-OS)/util/no_os_mutex.c
```
Comment (amiclaus) [inline projects/eval-adxl313z/src.mk:39]: ditto

File: projects/eval-adxl355-pmdz/src.mk [hunk 29]
```diff
@@ -25,7 +25,8 @@ INCS += $(INCLUDE)/no_os_delay.h     \
 		$(INCLUDE)/no_os_util.h \
 		$(INCLUDE)/no_os_units.h \
 		$(INCLUDE)/no_os_init.h \
-		$(INCLUDE)/no_os_alloc.h
+		$(INCLUDE)/no_os_alloc.h \
+        $(INCLUDE)/no_os_mutex.h
```
Comment (amiclaus) [inline projects/eval-adxl355-pmdz/src.mk:29]: ditto

File: projects/eval-adxl355-pmdz/src.mk [hunk 30]
```diff
@@ -36,7 +37,8 @@ SRCS += $(DRIVERS)/api/no_os_gpio.c \
 		$(DRIVERS)/api/no_os_uart.c \
 		$(NO-OS)/util/no_os_list.c \
 		$(NO-OS)/util/no_os_util.c \
-		$(NO-OS)/util/no_os_alloc.c
+		$(NO-OS)/util/no_os_alloc.c \
+        $(NO-OS)/util/no_os_mutex.c
```
Comment (amiclaus) [inline projects/eval-adxl355-pmdz/src.mk:41]: ditto.

File: projects/max31855/src.mk [hunk 31]
```diff
@@ -22,7 +22,8 @@ INCS += $(INCLUDE)/no_os_delay.h     \
 		$(INCLUDE)/no_os_uart.h      \
 		$(INCLUDE)/no_os_lf256fifo.h \
 		$(INCLUDE)/no_os_util.h 	\
-		$(INCLUDE)/no_os_units.h 
+		$(INCLUDE)/no_os_units.h	\
+        $(INCLUDE)/no_os_mutex.h
```
Comment (amiclaus) [inline projects/max31855/src.mk:26]: ditto.

File: projects/max31855/src.mk [hunk 32]
```diff
@@ -31,7 +32,8 @@ SRCS += $(DRIVERS)/api/no_os_gpio.c \
 		$(DRIVERS)/api/no_os_uart.c \
 		$(NO-OS)/util/no_os_list.c \
 		$(NO-OS)/util/no_os_util.c \
-		$(NO-OS)/util/no_os_alloc.c
+		$(NO-OS)/util/no_os_alloc.c \
+        $(NO-OS)/util/no_os_mutex.c
```
Comment (amiclaus) [inline projects/max31855/src.mk:36]: ditto.

**Review by amiclaus — COMMENTED** [review]
some comments on my side.

File: tools/scripts/freertos.mk (hunk 3 above)
Comment (rbudai98) [reply tools/scripts/freertos.mk:31]: As per discussed it will be solved in a different PR

File: projects/adxrs290-pmdz/src.mk (hunk 24 above)
Comment (rbudai98) [reply projects/adxrs290-pmdz/src.mk:36]: but it was odd :))

**Comment by rbudai98** [comment]
v3:
* addressed PR comments

File: include/no_os_mdio.h (hunk 7 above)
Comment (rbudai98) [reply include/no_os_mdio.h]: Modified i2c and spi api to support bus descriptor, will add there

**Comment by rbudai98** [comment]
V4:
* removed commit 34ebcab, hence it was added in an other PR
* modified api files according to recent changes in master
* i2c and spi have bus that allocates mutex automatically, while other peripherals have individual mutexes

File: drivers/api/no_os_eeprom.c [hunk 33]
```diff
@@ -41,6 +41,12 @@
 #include <stdlib.h>
 #include "no_os_eeprom.h"
 #include "no_os_error.h"
+#include "no_os_mutex.h"
+
+/**
+ * @brief - EEPROM mutex
+*/
+static void *eeprom_mutex;
```
Comment (buha) [inline drivers/api/no_os_eeprom.c:49]: i'm not sure i understand the role of this global

Comment (buha) [reply drivers/api/no_os_eeprom.c:49]: (same comment for all the other files)

Comment (rbudai98) [reply drivers/api/no_os_eeprom.c:49]: This pointer is initialized only once to point to the allocated mutex. Otherwise is just a null pointer.

Comment (rbudai98) [reply drivers/api/no_os_eeprom.c:49]: Similar to this, but having only one element:
https://github.com/analogdevicesinc/no-OS/blob/dea741fba32637889047a3eba778565f3e3ecfe3/drivers/api/no_os_i2c.c#L50

Comment (rbudai98) [reply drivers/api/no_os_eeprom.c:49]: Will change it to array

**Comment by rbudai98** [comment]
v5:
* changed mutex pointers to array of pointers and added header defffinitions for max number of peripherals
* drivers/api/spi: transfer return value corrected
* include/no_os_mutex.h: added warning to header functions

File: drivers/api/no_os_eeprom.c [hunk 34]
```diff
@@ -64,6 +70,9 @@ int32_t no_os_eeprom_init(struct no_os_eeprom_desc **desc,
 		return ret;
 
 	(*desc)->platform_ops = param->platform_ops;
+	(*desc)->mutex = eeprom_mutex_table[param->device_id];
+
+	no_os_mutex_init((*desc)->mutex);
```
Comment (buha) [inline drivers/api/no_os_eeprom.c:75]: Do these inits actually work ?
I think they should be `no_os_mutex_init(&(*desc)->mutex);` (*address of a pointer*)

File: include/no_os_eeprom.h [hunk 35]
```diff
@@ -50,6 +50,8 @@
 /********************** Macros and Constants Definitions **********************/
 /******************************************************************************/
 
+#define EEPROM_MAX_NUMBER 4
```
Comment (buha) [inline include/no_os_eeprom.h:53]: For Pwm/Timer/Uart it's pretty clear, they are hw peripherals that have a "device id" and having a static table with mutexes for each makes sense. But in the case of EEPROM and MDIO, it's not clear to me how to handle them. I would propose to remove the eeprom/mdio changes from this PR, think them through and add them later, separately. For me t's not even clear what an EEPROM device ID is, and that ID is used as a index in the mutex table. So I think we need to sort that out in the team first.

**Comment by rbudai98** [comment]
v6:
* added dereferencing to the mutex initialization
* removed EEPROM and MDIO mutex implementation
* added slave counter and counter logic to spi and i2c bus descriptors

File: include/no_os_eeprom.h (hunk 35 above)
Comment (rbudai98) [reply include/no_os_eeprom.h]: removed

File: drivers/api/no_os_eeprom.c (hunk 34 above)
Comment (rbudai98) [reply drivers/api/no_os_eeprom.c]: added pointer reference

File: drivers/api/no_os_spi.c [hunk 36]
```diff
@@ -138,11 +140,15 @@ void no_os_spibus_remove(uint32_t bus_number)
 	struct no_os_spibus_desc *spi_bus_desriptor = (struct no_os_spibus_desc *)
 			spi_table[bus_number];
 
-	no_os_mutex_remove(spi_bus_desriptor->mutex);
+	spi_bus_desriptor->slave_number = spi_bus_desriptor->slave_number - 1;
```
Comment (buha) [inline drivers/api/no_os_spi.c:143]: you can simplify by doing `spi_bus_desriptor->slave_number--;`

Comment (buha) [reply drivers/api/no_os_spi.c:143]: also, since we're here, you should only do `slave_number--` if `slave_number > 0`, otherwise you can end up with a wrap around

Comment (buha) [reply drivers/api/no_os_spi.c:143]: these comments are valid both for spi and i2c

**Comment by buha** [comment]
One more thing, I think you need to changes this in no_os_spibus_remove():

```
no_os_spibus_remove(desc->device_id);
```

to this:
```
no_os_spibus_remove(desc->bus->device_id);
```

https://github.com/analogdevicesinc/no-OS/blob/23a85b3c04b13b5a67f0e2cff88b40f9745a4b46/drivers/api/no_os_spi.c#L124

Valid also for i2c.

**Comment by rbudai98** [comment]
v7:
* applied PR suggestions

**Review by buha — APPROVED** [review]

File: drivers/api/no_os_i2c.c [hunk 37]
```diff
@@ -137,11 +139,16 @@ void no_os_i2cbus_remove(uint32_t bus_number)
 	struct no_os_i2cbus_desc *i2c_bus_desriptor = (struct no_os_i2cbus_desc *)
 			i2c_table[bus_number];
 
-	no_os_mutex_remove(i2c_bus_desriptor->mutex);
+	if (i2c_bus_desriptor->slave_number > 0)
```
Comment (buha) [inline drivers/api/no_os_i2c.c:142]: rename `desriptor` -> `descriptor` or delete it altogether - it makes for an unnecessarily long name, it could just as well be a variable named `bus`

**Review by buha — APPROVED** [review]
looks good now, i am merging
